### PR TITLE
Spider: find more URLs in headers

### PR DESCRIPTION
**Pull request by psiinon**
Based on the Google Security Crawl Maze header tests.

Signed-off-by: Simon Bennetts <psiinon@gmail.com>

### CODE REVIEW

File: zap/src/main/java/org/zaproxy/zap/spider/parser/SpiderHtmlParser.java [hunk 1]
```diff
@@ -39,8 +39,8 @@
  */
 public class SpiderHtmlParser extends SpiderParser {
 
-    /** The Constant urlPattern defining the pattern for a meta url. */
-    private static final Pattern urlPattern =
+    /** The Constant URL_PATTERN defining the pattern for a meta url. */
+    public static final Pattern URL_PATTERN =
```
Comment (thc202) [inline zap/src/main/java/org/zaproxy/zap/spider/parser/SpiderHtmlParser.java:43]: `static final`

File: zap/src/main/java/org/zaproxy/zap/spider/parser/SpiderHttpHeaderParser.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+ * limitations under the License.
+ */
+package org.zaproxy.zap.spider.parser;
+
+import java.util.regex.Matcher;
+import net.htmlparser.jericho.Source;
+import org.parosproxy.paros.network.HttpHeader;
+import org.parosproxy.paros.network.HttpMessage;
+
+/** The Class SpiderHttpHeaderParser is used for parsing of HTTP headers that can include URLs. */
```
Comment (thc202) [inline zap/src/main/java/org/zaproxy/zap/spider/parser/SpiderHttpHeaderParser.java:27]: `@see SpiderRedirectParser`

File: zap/src/test/java/org/zaproxy/zap/spider/parser/SpiderHttpHeaderParserUnitTest.java [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+    @Test
+    void shouldExtractUrlsFromLinkHeader() {
+        // Given
+        String url1 = "http://example.com/link1";
+        String url2 = "/link2";
+        HttpMessage msg = createMessageWithStatusCode(HttpStatusCode.FOUND);
+        msg.getResponseHeader()
+                .addHeader(
+                        HttpHeader.LINK,
+                        "<" + url1 + ">; param1=value1; param2=\"value2\";<" + url2 + ">");
```
Comment (thc202) [inline zap/src/test/java/org/zaproxy/zap/spider/parser/SpiderHttpHeaderParserUnitTest.java:146]: Could you add a test with no `<`, `>`, and without one and the other?

File: zap/src/test/java/org/zaproxy/zap/spider/parser/SpiderHttpHeaderParserUnitTest.java [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+        SpiderHttpHeaderParser headerParser = new SpiderHttpHeaderParser();
+        TestSpiderParserListener listener = createAndAddTestSpiderParserListener(headerParser);
+        // When
+        boolean parsed = headerParser.parseResource(msg, null, BASE_DEPTH);
+        // Then
+        assertThat(parsed, is(equalTo(false)));
+        assertThat(listener.getUrlsFound(), contains("http://example.com" + url));
+    }
+
+    private static HttpMessage createMessageWithStatusCode(int statusCode) {
```
Comment (thc202) [inline zap/src/test/java/org/zaproxy/zap/spider/parser/SpiderHttpHeaderParserUnitTest.java:186]: `createMessage()`, the status is not relevant.

**Comment by psiinon** [comment]
Hopefully addressed all of those :)

**Review by thc202 — APPROVED** [review]

**Comment by thc202** [comment]
Thank you!

**Review by kingthorin — APPROVED** [review]
Nice work

**Pull request merged**